src/Cassandra/Connections/BaseEndPointResolver.cs

```diff
@@ -37,8 +37,13 @@ internal abstract class BaseEndPointResolver : IEndPointResolver
         /// <summary>
         /// This method makes sure that there are no concurrent refresh operations.
         /// </summary>
-        protected async Task SafeRefreshAsync(Func<Task> refreshFunc)
+        protected async Task SafeRefreshIfNeededAsync(Func<bool> refreshNeeded, Func<Task> refreshFunc)
         {
+            if (!refreshNeeded())
```

Comment:
What's the reason for moving the condition to handler?

Reply:
I had a unit test fail where the refresh was being called more than once in a scenario with some concurrency. I realized that there was a race condition here since this check wasn't being made within the critical section so consider a scenario with threads A and B:

Thread A passes the check, enters SafeRefreshAsync
Thread B passes the check, enters SafeRefreshAsync
Thread A enters critical section, launches a new task and sets `_currentTask`
`_currentTask` finishes
Thread B enters critical section, checks that `_currentTask` has finished so it launches a new one.

As I typed this I'm realizing that maybe the condition could stay where it was before and what actually fixed this is the new `!object.ReferenceEquals(newTask, task))` check that I also added at the same time... I'll try this really quick.

Reply:
Wait, no then we would have an issue like this:

Thread A passes the check, enters SafeRefreshAsync
Thread B passes the check, enters SafeRefreshAsync
Thread A checks that there is no running task, proceeds.
Thread A enters critical section, launches a new task and sets _currentTask
_currentTask finishes
Thread B checks that there is no running task, proceeds.
Thread B enters critical section, launches a new task.

The check inside the critical section would allow B to detect that refresh is no longer needed even if it didn't detect that another thread launched a task at the same time.

Reply:
makes sense, thanks for the explanation